R/pivot_tables.R

```diff
@@ -357,3 +363,50 @@ ColocalizationScoresToAssay.Seurat <- function (
 
   return(object)
 }
+
+#' Utility function to fetch spatial metrics from
+#' a CellGraphAssay(5) object and convert them to a matrix
+#'
+#' @noRd
+#'
+.create_spatial_metric_matrix <- function (
+  object,
+  values_from,
+  metric,
+  ...
+) {
+  # fetch polarization scores
+  spatial_metric_table <- slot(object, name = metric)
+  if (is.null(spatial_metric_table)) {
+    abort(glue("'{metric}' scores are missing from '{class(object)}'"))
+  }
+
+  # Validate input
+  values_from <- match.arg(values_from, choices = switch(
+    metric,
+    "polarization" = c("morans_z", "morans_i"),
+    "colocalization" = c("pearson_z", "pearson", "pearson_mean",
+                         "jaccard_mean", "jaccard", "jaccard_z")
+  ))
+
+
+  pivot_assay_func <- switch(
+    metric,
+    "polarization" = PolarizationScoresToAssay,
+    "colocalization" = ColocalizationScoresToAssay
+  )
+  spatial_metric_wide_format <- pivot_assay_func(spatial_metric_table, values_from, ...)
+
+  # Make sure that the components match
+  # Create an empty matrix (all 0's)
+  tofillMat <- matrix(data = 0,
+                      nrow = nrow(spatial_metric_wide_format),
+                      ncol = ncol(object),
+                      dimnames = list(rownames(spatial_metric_wide_format), colnames(object)))
+
+  # Fill matrix where it overlaps
+  # Any missing columns will be kept as 0's
+  tofillMat[, colnames(spatial_metric_wide_format)] <- spatial_metric_wide_format
```

Comment:
Why is this necessary when `pivot_wider` is used? It should fill in with zeros anyway?

Reply:
The function `PolarizationScoresToAssay.data.frame` fills with 0's using `pivot_wider` and then again explicitly: 

```
# Cast data.frame to wide format
  pol_scores_wide_format <- object %>%
    pivot_wider(id_cols = "marker",
                names_from = "component",
                values_from = all_of(values_from),
                values_fill = 0) %>%
    data.frame(row.names = 1) %>%
    as.matrix()
  # Replace missing values with 0
  pol_scores_wide_format[is.na(pol_scores_wide_format)] <- 0
```

Perhaps we should remove the explicit replacement up there.

Reply:
Right, must have forgotten to remove that!